File: src/gateway/handlers/heartbeat.rs
use axum::{extract::Path, http::StatusCode, response::IntoResponse, Json};

use super::super::types::{HeartbeatStatusItem, HeartbeatStatusResponse};
use super::super::utils::{not_found_response, validate_or_return};

/// `GET /api/heartbeat/status` — list heartbeat status for all agents.
pub(crate) async fn api_heartbeat_status_all() -> impl IntoResponse {
    let mut statuses = Vec::new();

    // Use shared helper to iterate agent directories
    let agent_entries = crate::gateway::utils::iter_agents_dir().await;
    for entry in agent_entries {
        if let Some(status) = crate::scheduler::load_heartbeat_status(&entry.agent_id).await {
            statuses.push(heartbeat_status_to_item(&status));
        }
    }

    Json(HeartbeatStatusResponse { agents: statuses })
}

/// `GET /api/heartbeat/status/:agent_id` — heartbeat for one agent.
pub(crate) async fn api_heartbeat_status_one(Path(agent_id): Path<String>) -> impl IntoResponse {
    validate_or_return!(&agent_id);

    match crate::scheduler::load_heartbeat_status(&agent_id).await {
        Some(status) => {
            let item = heartbeat_status_to_item(&status);
            (StatusCode::OK, Json(item)).into_response()
        }
        None => not_found_response(agent_id),
    }
}

pub(crate) fn heartbeat_status_to_item(
    s: &crate::scheduler::HeartbeatStatus,
) -> HeartbeatStatusItem {
    let health = match &s.health {
        crate::scheduler::HeartbeatHealth::OK => "OK".to_string(),
        crate::scheduler::HeartbeatHealth::MISSED => "MISSED".to_string(),
        crate::scheduler::HeartbeatHealth::ERROR(e) => format!("ERROR: {e}"),
    };
    HeartbeatStatusItem {
        agent_id: s.agent_id.clone(),
        enabled: s.enabled,
        health,
        last_tick: s.last_tick,
        next_tick: s.next_tick,
        interval_secs: s.interval_secs,
        message_preview: s.message_preview.clone(),
        latest_session: s.latest_session.clone(),
    }
}

File: web/features/agents/api/heartbeat.ts
/**
 * Heartbeat API - Agent heartbeat status operations
 * Endpoints from src/gateway/handlers/heartbeat.rs
 */

import { fetchApi } from '@/shared/api/client';
import type { HeartbeatStatusResponse, HeartbeatStatusItem } from '@/src/lib/bindings';
import type { HeartbeatStatusData } from '../types';

const API_BASE = '/api/heartbeat';

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Transform raw heartbeat status to frontend format
 */
function transformHeartbeatStatus(raw: HeartbeatStatusItem): HeartbeatStatusData {
  return {
    agentId: raw.agent_id,
    enabled: raw.enabled,
    health: raw.health as 'OK' | 'MISSED',
    lastTick: raw.last_tick ? new Date(Number(raw.last_tick) * 1000).toISOString() : null,
    nextTick: raw.next_tick ? new Date(Number(raw.next_tick) * 1000).toISOString() : null,
    intervalSecs: raw.interval_secs ? Number(raw.interval_secs) : null,
    messagePreview: raw.message_preview ?? undefined,
    latestSession: typeof raw.latest_session === 'string' && raw.latest_session
      ? { id: raw.latest_session }
      : undefined,
  };
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Get heartbeat status for a single agent
 * GET /api/heartbeat/status/:agent_id
 */
export async function getAgentHeartbeat(agentId: string): Promise<HeartbeatStatusData> {
  const response = await fetchApi<HeartbeatStatusItem>(
    `${API_BASE}/status/${encodeURIComponent(agentId)}`
  );
  return transformHeartbeatStatus(response);
}

/**
 * Get heartbeat status for all agents
 * GET /api/heartbeat/status
 */
export async function getAllAgentsHeartbeat(): Promise<HeartbeatStatusData[]> {
  const response = await fetchApi<HeartbeatStatusResponse>(`${API_BASE}/status`);
  return response.agents.map(transformHeartbeatStatus);
}

/**
 * Get heartbeat status for multiple specific agents
 * Helper function to batch fetch heartbeats
 */
export async function getAgentsHeartbeat(
  agentIds: string[]
): Promise<Map<string, HeartbeatStatusData>> {
  const allHeartbeats = await getAllAgentsHeartbeat();
  const result = new Map<string, HeartbeatStatusData>();

  for (const hb of allHeartbeats) {
    if (agentIds.includes(hb.agentId)) {
      result.set(hb.agentId, hb);
    }
  }

  return result;
}

/**
 * Check if an agent's heartbeat is healthy
 */
export function isHeartbeatHealthy(status: HeartbeatStatusData): boolean {
  return status.enabled && status.health === 'OK';
}

/**
 * Check if an agent's heartbeat has issues
 */
export function isHeartbeatMissed(status: HeartbeatStatusData): boolean {
  return status.enabled && status.health === 'MISSED';
}

/**
 * Get time until next heartbeat
 * Returns null if heartbeat is not enabled or next tick unknown
 */
export function getTimeToNextHeartbeat(status: HeartbeatStatusData): number | null {
  if (!status.enabled || !status.nextTick) {
    return null;
  }

  const nextTick = new Date(status.nextTick).getTime();
  const now = Date.now();
  const diff = nextTick - now;

  return Math.max(0, diff);
}

/**
 * Format heartbeat interval for display
 */
export function formatHeartbeatInterval(seconds: number | null): string {
  if (seconds === null || seconds === undefined) {
    return 'Disabled';
  }

  if (seconds < 60) {
    return `${seconds}s`;
  }

  if (seconds < 3600) {
    const minutes = Math.floor(seconds / 60);
    return `${minutes}m`;
  }

  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);

  if (minutes === 0) {
    return `${hours}h`;
  }

  return `${hours}h ${minutes}m`;
}
